### ISSUE TITLE
pytest 2.8.1: wrong error type is caught in _pytest/python.py in def _idval(val, argname, idx, idfn)

### ISSUE THREAD

**Issue by sergiy-kozak** [pytest-dev/pytest#1099]
In function _idval(val, argname, idx, idfn) of mentioned source file next code is present:
```python
elif _PY2 and isinstance(val, unicode):
    # special case for python 2: if a unicode string is
    # convertible to ascii, return it as an str() object instead
    try:
        return str(val)
    except UnicodeDecodeError:
        # fallthrough
        pass
return str(argname)+str(idx)
```
In case val variable is unicode value, that contains characters with ordinal > 128, str(val) fails with UnicodeEncodeError, which one would like to catch and fall back to autogenerated parameter id value, like argname1, argname2 etc. The current _idval code causes to the collection time exception, for example `UnicodeEncodeError: 'ascii' codec can't encode characters in position 3-5: ordinal not in range(128)` due to wrong error type being caught. To fix the problem, replace line
```python
    except UnicodeDecodeError:
```
with
```python
    except UnicodeEncodeError:
```

Environment:
platform linux2 -- Python 2.7.9, pytest-2.8.1, py-1.4.30, pluggy-0.3.1 -- /home/myuser/.virtualenvs/test-env/bin/python

**Comment by sergiy-kozak** [comment]
Similar problem exists in the following function implementation (in _pytest/python.py):
```python
    def _escape_bytes(val):
        """
        In py2 bytes and str are the same, so return it unchanged if it
        is a full ascii string, otherwise escape it into its binary form.
        """
        try:
            return val.encode('ascii')
        except UnicodeDecodeError:
            return val.encode('string-escape')
```
which may be fixed by replacing `except UnicodeDecodeError:` with `except UnicodeEncodeError:`

**Comment by The-Compiler** [comment]
It seems at least the first case was fixed in #1092

**Comment by nicoddemus** [comment]
@sergiy-kozak could you try with latest `master` please? This should be fixed there. You can install it directly into a virtual environment:

```
pip install git+https://github.com/pytest-dev/pytest.git
```

**Comment by sergiy-kozak** [comment]
I'll give it a try and report here.

**Comment by sergiy-kozak** [comment]
I tested with the latest dev pytest code. I confirm the fix in #1092 is sufficiently solving the issue, hence #1099 can be closed as the duplicate. Thanks everyone for replying on this issue.

**Comment by nicoddemus** [comment]
Thanks! :smile:
